audit/base/base.go

```diff
@@ -0,0 +1,556 @@
+to send to another audit server or drop them. You can reset the client by using the Reset() method with
+a new connection.
+
+The audit client is designed to be used as a singleton. This means that you should only have one
+at any time per application.
+
+You can adjust the settings for the audit client by using the WithSettings() option.
+*/
+package base
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"runtime"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/microsoft/go-otel-audit/audit/conn"
+	"github.com/microsoft/go-otel-audit/audit/internal/version"
+	"github.com/microsoft/go-otel-audit/audit/msgs"
+)
+
+// ctxBack is used as a background context when we cannot use a context passed
+// to a method (such as when sending a heartbeat). Putting it here avoids an
+// allocation every time we need a background context in this manner.
+var ctxBack = context.Background()
+
+// Default values for the audit client.
+const (
+	// MaxQueueSize is the maximum number of audit records that can be queued.
+	MaxQueueSize = 2048
+)
+
+// ErrorCategory represents the category of an error.
+//
+//go:generate stringer -type=ErrorCategory
+type ErrorCategory uint8
+
+const (
+	// UnknownErrorCategory is an unknown error category. This indicates a bug in the audit client.
+	UnknownErrorCategory ErrorCategory = 0
+	// ErrValidation is an error that occurred during validation of an audit record.
+	// This means the audit record is invalid and was not sent.
+	ErrValidation ErrorCategory = 1
+	// ErrConnection is an error that occurred while connecting to the audit server.
+	// This means the audit client is in an unrecoverable state. Getting this error means that every
+	// other call to the client will fail with the same message.
+	ErrConnection ErrorCategory = 2
+	// ErrQueueFull is an error that occurred because the queue is full.
+	ErrQueueFull ErrorCategory = 3
+	// ErrTimeout is an error that occurred because the send timed out.
+	ErrTimeout ErrorCategory = 4
+)
+
+// Error represents an error that occurred in this package. All errors returned by this package
+// are of type Error.
+type Error struct {
+	// Err is the underlying error.
+	Err error
+	// Category is the category of the error.
+	Category ErrorCategory
+}
+
+// Error returns the error message.
+func (e Error) Error() string {
+	return fmt.Sprintf("%s: %s", e.Category, e.Err.Error())
+}
+
+// Unwrap returns the underlying error.
+func (e Error) Unwrap() error {
+	return e.Err
+}
+
+// IsUnrecoverable returns true if the error is unrecoverable. This only works on
+// errors of type Error (from this package). If not Error, this returns false. Unrecoverable errors
+// mean that the audit client should not attempt to send any more audit records as the
+// client is in an unrecoverable state (e.g. the connection is dead).
+func IsUnrecoverable(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if e, ok := err.(Error); ok {
```

Comment:
That seems unlikely as a use case, but the argument is still valid on:  why aren't you using errors.As().  And my excuse is really muscle memory, I tend to forget errors.Is() and As() because I didn't use them forever.

I've switched them all over to just being string errors that are wrapped into an error or used directly. Now you can just use errors.Is() and As().  I did keep IsUnrecoverable(), because we might have more categories that come up that fit this.